addons/format_lame.c

```diff
@@ -0,0 +1,541 @@
+
+	res = lame_encode_flush(p->lgfp, last_encoder_buffer, sizeof(last_encoder_buffer));
+	if (res < 0) {
+		ast_log(LOG_WARNING, "LAME encode last returned error %d\n", res);
+		lame_close(p->lgfp);
+		return -1;
+	} else if (res > 0) {
+		wres = fwrite(last_encoder_buffer, 1, res, fs->f);
+		if (wres != res) {
+			ast_log(LOG_WARNING, "Bad write (%d/%d): %s\n", wres, res, strerror(errno));
+			lame_close(p->lgfp);
+			return -1;
+		} else {
+			ast_debug(3, "LAME encode flushed %d bytes", wres);
+		}
+	}
+
+	return lame_close(p->lgfp);
+}
+
+static int mp3lame_decoder_init(struct ast_filestream *fs)
+{
+	int size;
+	struct mp3lame_private *p = fs->_private;
+
+	p->hgfp = hip_decode_init();
+
+	/* Get filesize in bytes */
+	fseek(fs->f, 0, SEEK_END);
+	size = ftell(fs->f);
+	fseek(fs->f, 0, SEEK_SET);
+
+	return size;
+}
+
+static int mp3lame_decoder_deinit(struct mp3lame_private *p)
+{
+	return hip_decode_exit(p->hgfp);
+}
+
+static int mp3lame_hip_decode(hip_global_flags * handle, int input_size, unsigned char * input, short int * output, int sample_rate)
+{
+	int samples;
+	mp3data_struct headers;
+
+	samples = hip_decode_headers(handle, input, input_size, output, NULL, &headers);
+
+	if (headers.header_parsed == 1) {
+		ast_debug(3, "LAME decoder found MP3 headers:\nchannels=%d, samplerate=%d, bitrate=%d, framesize=%d, mode=[%d:%d]",
+			headers.stereo,
+			headers.samplerate,
+			headers.bitrate,
+			headers.framesize,
+			headers.mode,
+			headers.mode_ext
+		);
+		if (headers.stereo < 1) {
+			ast_debug(3, "LAME decoder no audio channels");
+			return -1;
+		} else if (headers.stereo > 1) {
+			ast_log(LOG_ERROR, "LAME decoder invalid number of channels: %d, only mono is acceptable", headers.stereo);
+			return -2;
+		}
+		if (headers.samplerate != sample_rate) {
+			ast_log(LOG_ERROR, "LAME decoder invalid sampling rate: %d, expected %d", headers.samplerate, sample_rate);
+			return -3;
+		}
+	}
+
+	return samples;
+}
+
+static int mp3lame_file_read(struct ast_filestream *s, int sample_rate)
+{
+	int rres;
+	struct mp3lame_private *p = s->_private;
+	unsigned char dinput[p->total_bytes_compressed];
+
+	rres = fread(dinput, 1, p->total_bytes_compressed, s->f);
+	if(rres != p->total_bytes_compressed){
+		ast_log(LOG_ERROR, "Short read (%d/%d)", rres, p->total_bytes_compressed);
+		return -1;
+	}
+
+	ast_debug(3, "LAME decoder input: %d bytes", rres);
+	p->dsamples = mp3lame_hip_decode(p->hgfp, rres, dinput, p->doutput, sample_rate);
+	while (p->dsamples == 0) {
+		p->dsamples = mp3lame_hip_decode(p->hgfp, 0, NULL, p->doutput, sample_rate);
+	}
+	ast_debug(3, "LAME decoder got %d output samples", p->dsamples);
+
+	return p->dsamples * BITS;
+}
+
+static int mp3lame_open(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
+	p->encoder_counter = 0;
+	p->decoder_counter = 0;
+
+	return 0;
+}
+
+static void mp3lame_close(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
```

Comment:
Does the decoder need to be deinit here as well?

Reply:
No. deinit is in mp3lame_read(). The entire mp3 -> slin decoding is done when asterisk requests the first set of slin audio samples. The decoded audio is stored in the mp3lame_private->doutput buffer.

See: https://github.com/asterisk/asterisk/pull/665#issuecomment-2046534183